pkg/config/nodetreemodel/config.go

```diff
@@ -50,6 +50,11 @@ type ntmConfig struct {
 	root   Node
 	noimpl notImplementedMethods
 
+	// defaultBuilder is used during Setup to construct the default source tree,
+	// it will be nil once the config is marked as "Ready" for use
+	defaultBuilder    map[string]interface{}
```

Comment:
Why aggregate the defaults in a map instead of directly setting them in the `defaultSourceRoot` tree ?

Would it not be the same to do:
```
func (c *ntmConfig) SetDefault(key string, default interface{}) {
    keyParts = strings.Split(strings.ToLower(key), ".")
    c.SetAt(c.defaultSourceRoot, keyParts, default)
}
```

Reply:
My original hope was that we could construct the Nodes in the default tree all at once, in a functional style, rather than needing to mutate existing nodes by inserting children into them after they already have been constructed. This would remove the possibility of changing the shape of the default tree after it got built. However, the `isReady()` check pretty much accomplishes the same thing, so I don't think it's worth the trouble.

Reply:
Switched to your suggestion. There is one important difference however: both `SetDefault` and `BindEnv` can introduce a known key. If code does this:

```
cfg.SetDefault("a.b.c", myValue)
cfg.BindEnv("a.b.c", "DD_MY_ENV_VAR")
```

This will override the default `myValue` with nil, which is something that the old code handled. I left a TODO for now, we can fix that in a follow-up PR.